Benchmark against Sentry submodule

lafrenierejm opened the pull request:
This is intended to address concerns about reproducibility raised in sirwart/ripsecrets#25. As [I commented](https://github.com/sirwart/ripsecrets/pull/25#issuecomment-1236364673) on that issue:
> Git submodules natively use specific SHAs, so bit-for-bit reproducibility of the test data would be handled for us.

lafrenierejm committed:
Run Criterion benchmarks against Sentry submodule

lafrenierejm committed:
Reduce Criterion sample size from 150 to 30

lafrenierejm committed:
Include gnuplot in Nix dev dependencies

gnuplot is used by Criterion to generate its visualizations.

lafrenierejm committed:
Document Criterion's output path

sts10 commented:
Neat! 

Curious why you advise users to run `cargo bench >/dev/null`? While the terminal output of `cargo bench` isn't as detailed as the HTML report, I think it's still useful. Here's what mine was: 
```text
Find secrets in getsentry/sentry/find_secrets function                        
                        time:   [7.3808 ms 7.4233 ms 7.4710 ms]
Found 1 outliers among 30 measurements (3.33%)
  1 (3.33%) low mild
```

That terminal output does also print this warning: 
```text
WARNING: HTML report generation will become a non-default optional feature in Criterion.rs 0.4.0.
This feature is being moved to cargo-criterion (https://github.com/bheisler/cargo-criterion) and will be optional in a future version of Criterion.rs. To silence this warning, either switch to cargo-criterion or enable the 'html_reports' feature in your Cargo.toml.
```

Wonder if we can address that before merging into sirwart's repo (I was lazy!).

lafrenierejm commented:
I described the motivation for redirecting to `/dev/null` in the relevant commit's message:

> This prevents the throughput of one's terminal emulator from affecting the benchmark results.

That said, I'd be totally fine with dropping the redirection from the readme. I suspect that in the vast majority of cases any performance regression  testing would be performed with the same stdout device and therefore not have a material impact on relative performance.

I'll try to get around to dropping that commit within the next day.

lafrenierejm commented:
> Wonder if we can address that before merging into sirwart's repo (I was lazy!).

Yeah, I think it would be worth addressing that sooner rather than later. I'm not really familiar with Cargo, if I figure out a good config I'll raise a separate PR.

sts10 commented:
Sorry for not investigating the commit messages.

> This prevents the throughput of one's terminal emulator from affecting the benchmark results.

I'd be pretty surprised if this affected Criterion's measurements? My basic conceptual understanding of a good bench-marking tool is that the benchmarks are run in a sort of "black box" to eliminate as many variables as possible. I'd assume this includes things like printing the results to the terminal, just as it (hopefully) includes the time it takes to generate the HTML of the report. 

Lastly, I suggest that, while it's admirable to try to equalize benchmark times across different users' machines, that's ultimately not really possible given different CPUs, etc. As I understand it, the power of benchmarking comes from one user running the bench on two different versions (say, the existing release versus proposed changes) and comparing the _relative_ times.

lafrenierejm commented:
> As I understand it, the power of benchmarking comes from one user running the bench on two different versions (say, the existing release versus proposed changes) and comparing the relative times.

Agreed. When I added the `/dev/null` redirection recommendation I was mostly thinking forward to running benchmarks as part of CI, but _if_ we encounter issues in CI then I think it would be better to solve the problems in that specific context rather than make a general recommendation.

I have dropped the commit in question from this branch.

lafrenierejm commented:
>> Wonder if we can address [criterion's warnings under `cargo bench`] before merging into sirwart's repo (I was lazy!).

> Yeah, I think it would be worth addressing that sooner rather than later. I'm not really familiar with Cargo, if I figure out a good config I'll raise a separate PR.

This was raised as #2.
